web_form_banner/models/web_form_banner_rule.py

```diff
@@ -0,0 +1,287 @@
+# Copyright 2025 Quartile (https://www.quartile.co)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+import logging
+from functools import lru_cache
+from string import Template
+
+from dateutil import parser as dateparse
+from dateutil.relativedelta import relativedelta
+from lxml import etree
+from pytz import timezone
+
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+from odoo.tools import html_escape, safe_eval
```

Comment:
To avoid unnecessary diffs. Use `tools.safe_eval` for `time` and `datetime` in `_base_eval_ctx_static()`.
```suggestion
from odoo import _, api, fields, models, tools
from odoo.exceptions import ValidationError
from odoo.tools import html_escape
from odoo.tools.safe_eval import safe_eval
```